akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -473,6 +473,177 @@ private[akka] final case class Conflate[In, Out](seed: In ⇒ Out, aggregate: (O
   override def restart(): Conflate[In, Out] = copy()
 }
 
+private[akka] sealed abstract class AbstractBatch[In, Out](max: Long, costFn: In ⇒ Long, seed: In ⇒ Out,
+                                                           aggregate: (Out, In) ⇒ Out, val in: Inlet[In],
+                                                           val out: Outlet[Out]) extends GraphStage[FlowShape[In, Out]] {
+
+  override val shape: FlowShape[In, Out] = FlowShape.of(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var agg: Any = null
+    private var left: Long = max
+    private var pending: Any = null
+
+    private def flush(): Unit = {
+      push(out, agg.asInstanceOf[Out])
+      left = max
+      if (pending != null) {
+        val elem = pending.asInstanceOf[In]
+        agg = seed(elem)
+        left -= costFn(elem)
+        pending = null
+      } else {
+        agg = null
+      }
+    }
+
+    override def preStart() = pull(in)
+
+    setHandler(in, new InHandler {
+
+      override def onPush(): Unit = {
+        val elem = grab(in)
+        val cost = costFn(elem)
+        if (agg == null) {
+          left -= cost
+          agg = seed(elem)
+        } else if (left < cost) {
+          pending = elem
+        } else {
+          left -= cost
+          agg = aggregate(agg.asInstanceOf[Out], elem)
+        }
+
+        if (isAvailable(out)) flush()
+        if (pending == null) pull(in)
```

Comment:
That's true. but since I can still store a pending element, shouldn't I be eager about it?
Fetching the element at this state would indeed just store it as pending, but I think it might be considered as an optimization especially if the publisher may be slower. WDYT?